libs/cache_line/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+#![feature(const_trait_impl)]
+#![no_std]
+
+use core::{mem, ops};
+
+const LINE_SIZE: usize = 64;
+
+/// A cache line optimised array.
+///
+/// It spreads elements across as many cache lines as possible. For example if
+/// only the first four elements of a 256 byte array are being accessed, each
+/// element will have its own cache line.
+///
+/// Indexing is faster if the number of cache lines that the array spreads is a
+/// power of two.
+#[derive(Debug)]
+// Alignment must be the same as `LINE_SIZE`.
+#[repr(align(64))]
+pub struct CacheLineOptimisedArray<T, const N: usize> {
+    inner: [T; N],
+}
+
+impl<T, const N: usize> const From<[T; N]> for CacheLineOptimisedArray<T, N> {
+    fn from(value: [T; N]) -> Self {
+        Self { inner: value }
+    }
+}
```

Comment:
hmm, perhaps i'm misunderstanding this approach, but don't we actually want each element in the array to occupy a separate cache line? That would imply that each `T` needs to be 64-byte aligned in addition to the array itself. This also means that we wouldn't be able to directly construct the array from an existing `[T; N]` value unless the alignment of T was already 64 and the size of T was <= 64.

It seems like the current impl is more focused on putting the multiple elements in a single cache line, which is less beneficial than one element per cache line, at least for the preemption example where we want to avoid cache line thrashing and false sharing.

Reply:
No, what we want is to spread the elements across different cache lines as much as possible. So, for example, the array of preemption counters will still only take up 256 bytes, but in the majority of cases when we are only using 4 CPUs, each CPU preemption counter will have its own cache line. If you add more CPUs, they will be equally spread out across the cache lines, so CPUs 0 and 4 will be in cache line 1, CPUs 1 and 5 in line 2, CPUs 2 and 6 in line 3, and CPUs 3 and 7 in line 4.

Reply:
Right, but at least for things that should be CPU-local, we don't want data for one CPU to share a cache line with that of another CPU. It's possible that other categories of data may be okay with this sharded approach, but I can't see how it's beneficial for this specific case.

Basically, I would expect this to look more like `[CachePadded<T>; N]`, since that's what would actually reduce the bad false sharing (across CPUs) that we want to avoid.

As you know, we're now about to use CPU locals for this instead, but I think it would be good to go through other types that are frequently shared across CPUs and use the `crossbeam_utils::CachedPadded` type to wrap them. There are other places where we want to align data structures up to a cache line size, but the items shared across CPUs are the lowest hanging fruit therein.